DevTools/UserspaceEmulator/Emulator.cpp

```diff
@@ -1450,4 +1455,47 @@ int Emulator::virt$sched_setparam(int pid, FlatPtr user_addr)
     return syscall(SC_sched_setparam, pid, &user_param);
 }
 
+void thread_entry_point(Emulator* emu, FlatPtr entry, FlatPtr esp)
+{
+    dbg() << "Thread entry point, emu=" << emu << ", entry=" << entry;
+    emu->m_cpu.set_eip((FlatPtr)entry);
+    // FIXME: Don't assume u32.
+    emu->m_cpu.set_esp(shadow_wrap_as_initialized((u32)esp));
+    emu->exec();
```

Comment:
How is this expected to work? The thread will clobber the CPU registers used by the emulator on the main thread.
Secondary threads should either have a separate CPU, or the emulator should have a thread scheduler that periodically switches between threads (while preserves the register state of each thread.)

Reply:
Oh.
I seem to have mistaken it for a separate process...oops

Reply:
I can think of two things to do about this (just your suggestions, really), and I don't like either of them:
- keep a bunch of `SoftCPU`'s in a hashmap or something, one for each active thread
- write another scheduler, as the one in Kernel is a bit too specialised

Both of these sound terrible, but I'm leaning towards option A - alternatively, dropping that commit entirely.
Any suggestions or alternatives you can think of?

Reply:
Nevermind, putting a `SoftCPU` in a hashmap is annoying, that specific syscall can wait a little longer :shrug: